### PR TITLE
feat(wallet): add smart sessions ownable validator support

### PR DESCRIPTION
**Pull request by rplusq**
# Description

Adds support for Rhinestone's **OwnableValidator** to replace our custom MultiKeySigner, enabling us to use their audited validator instead of requiring our own security audit.

## ⚠️ Breaking Changes

**Cosigner public key format change:**
```rust
// Before
- public_key: format!("0x{}", hex::encode(public_key_der_hex)),

// After  
+ public_key: format!("0x{public_key_der_hex}"),
```

**Impact:** The create endpoint now returns the cosigner public key as hex string directly (not ascii-hex encoded). This matches the expected AppKit Wagmi Async permissions format but may affect other integrations.

**Migration:** Any consumers expecting the old ascii-hex format will need to update their parsing logic.

## Key changes:
- Auto-detects OwnableValidator vs MultiKeySigner format via `is_ownable_validator_format()`
- Handles different signature formats (concatenated for OwnableValidator, ABI-encoded for MultiKeySigner)
- Updates validation logic to match OwnableValidator contract constraints (max 32 owners, threshold ≤ owners)
- Adds proper error handling with detailed `PermissionContextError` variants

**Context:** Team discussion identified we can avoid audit costs by using Rhinestone's already-audited OwnableValidator since we don't currently need passkey support (EOA-only works fine).

## How Has This Been Tested?

- ✅ **9 validator tests** covering format detection and validation edge cases
- ✅ **12 prepare_calls tests** verifying signature format conversion  
- ✅ All existing tests pass - backward compatible with MultiKeySigner

```bash
cargo test --lib utils::validators
cargo test --lib handlers::wallet::prepare_calls
```

## Due Diligence

* [x] Breaking change
* [ ] Requires a documentation update  
* [x] Requires a e2e/integration test update

### CODE REVIEW

**Review by geekbrother — COMMENTED** [review]
Thanks a lot 👍 
Can you please check clippy and formating errors 🙏 
We have basic integration tests for sessions [here](https://github.com/reown-com/blockchain-api/blob/master/integration/sessions.test.ts). I believe we need to change the format here also to pass it.

File: src/handlers/wallet/prepare_calls.rs [hunk 1]
```diff
@@ -147,6 +172,27 @@ pub enum PrepareCallsError {
     InternalError(PrepareCallsInternalError),
 }
 
+#[derive(Error, Debug)]
+pub enum PermissionContextError {
```
Comment (geekbrother) [inline src/handlers/wallet/prepare_calls.rs:176]: Nice, thanks for this 👍

File: src/handlers/sessions/cosign.rs [hunk 2]
```diff
@@ -220,10 +230,27 @@ async fn handler_internal(
     let signature = wallet
         .sign_hash(H256::from(&keccak256(eip191_user_op_hash.clone())))
         .unwrap();
+
+    let message_hash = H256::from(&keccak256(eip191_user_op_hash.clone()));
```
Comment (tomiir) [inline src/handlers/sessions/cosign.rs:234]: curious: how come we'd tried this and it had failed?

File: src/handlers/wallet/prepare_calls.rs [hunk 3]
```diff
@@ -657,21 +741,37 @@ async fn get_dummy_signature(
...
+    } else {
+        // For MultiKeySigner: ABI-encode as array
+        let signature_array = signers
+            .into_iter()
+            .map(|t| match t {
+                SignerType::Ecdsa => DUMMY_ECDSA_SIGNATURE,
+                SignerType::Passkey => DUMMY_PASSKEY_SIGNATURE,
+            })
+            .collect::<Vec<_>>();
+        signature_array.abi_encode()
```
Comment (tomiir) [inline src/handlers/wallet/prepare_calls.rs:773]: should we remove this ?

File: src/handlers/sessions/create.rs [hunk 4]
```diff
@@ -114,7 +117,7 @@ async fn handler_internal(
         pci: pci.clone(),
         key: KeyItem {
             r#type: KeyType::Secp256k1,
-            public_key: format!("0x{}", hex::encode(public_key_der_hex)),
+            public_key: format!("0x{public_key_der_hex}"),
```
Comment (chris13524) [inline src/handlers/sessions/create.rs:120]: Why are you changing this?

Comment (chris13524) [inline src/handlers/sessions/create.rs:120]: This looks like it's a breaking change

File: src/handlers/wallet/prepare_calls.rs [hunk 5]
```diff
@@ -610,32 +665,61 @@ fn encode_enable_signature(
...
+#[derive(Clone)]
 enum SignerType {
     Ecdsa,
     Passkey,
 }
 
 fn decode_signers(data: Bytes) -> Result<Vec<SignerType>, PrepareCallsError> {
-    let mut data = data.into_iter();
-    let signer_count = data
+    // Try ABI decoding first (threshold + address array format) - OwnableValidator
```
Comment (chris13524) [inline src/handlers/wallet/prepare_calls.rs:675]: We should also have the validator's address here too to avoid sniffing

File: src/utils/validators.rs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+const MIN_ABI_ENCODED_TUPLE_LENGTH: usize = 64;
+const MAX_OWNERS_COUNT: usize = 32; // From OwnableValidator.MAX_OWNERS
+
+/// Check if data is abi.encode(threshold, owners) format - indicates Ownable
+/// Validator
+///
+/// Validates both format and reasonable limits to prevent DoS attacks.
+pub fn is_ownable_validator_format(data: &Bytes) -> bool {
+    if data.len() < MIN_ABI_ENCODED_TUPLE_LENGTH {
+        return false;
```
Comment (nopestack) [inline src/utils/validators.rs:16]: Seems like a good idea to return a Result from this function as the failures can be nuanced.

File: src/handlers/sessions/cosign.rs [hunk 7]
```diff
@@ -220,10 +230,27 @@ async fn handler_internal(
...
+        // The order is important: signatures must be in the same order as sorted owner
+        // addresses
+        let mut concatenated_signature = Vec::new();
+        concatenated_signature.extend_from_slice(&user_op.signature); // Frontend signature (first signer)
+        concatenated_signature.extend_from_slice(&packed_signature); // Backend signature (second signer)
+        Bytes::from(concatenated_signature)
+    } else {
+        // For MultiKeySigner: ABI encode the signatures
+        abi_encode_two_bytes_arrays(&packed_signature, &user_op.signature)
+    };
```
[the remark this answers is not in the document]
Comment (rplusq) [reply src/handlers/sessions/cosign.rs]: Added some comments explaining this. Long story short: the contract sorts signers after recovery

File: src/handlers/sessions/cosign.rs (hunk 2 above)
Comment (rplusq) [reply src/handlers/sessions/cosign.rs:232]: wdym?

File: src/handlers/wallet/prepare_calls.rs [hunk 8]
```diff
@@ -657,21 +741,37 @@ async fn get_dummy_signature(
...
+    let signers = decode_signers(validator_init_data.clone())?;
+
+    let signature = if is_ownable_validator_format(validator_init_data) {
+        // For OwnableValidator: concatenate signatures directly (abi.encodePacked
+        // style)
+        signers
+            .into_iter()
+            .flat_map(|t| match t {
+                SignerType::Ecdsa => DUMMY_ECDSA_SIGNATURE.to_vec(),
+                SignerType::Passkey => DUMMY_PASSKEY_SIGNATURE.to_vec(),
```
[the remark this answers is not in the document]
Comment (rplusq) [reply src/handlers/wallet/prepare_calls.rs:768]: I'm just leaving this for backwards compatiblity

File: src/utils/validators.rs (hunk 6 above)
Comment (rplusq) [reply src/utils/validators.rs]: ended up removing it, using this: https://github.com/reown-com/blockchain-api/blob/062dc472b1c57250e691f1caf20fbafd76bbbafa/src/utils/validators.rs#L7-L9

File: src/handlers/sessions/create.rs [hunk 9]
```diff
@@ -110,11 +113,23 @@ async fn handler_internal(
...
+    let public_key = match query_params.api_version {
+        Some(2) => {
+            // v2: Direct hex string with 0x prefix
+            format!("0x{public_key_der_hex}")
+        }
+        _ => {
+            // v1 (default): ASCII-hex encoded for backward compatibility
+            hex::encode(public_key_der_hex)
+        }
+    };
```
[the remark this answers is not in the document]
Comment (rplusq) [reply src/handlers/sessions/create.rs]: yep, re-added, mb
